widget/slider.go

```diff
@@ -0,0 +1,289 @@
+	track  *canvas.Rectangle
+	active *canvas.Rectangle
+	thumb  *canvas.Circle
+
+	objects []fyne.CanvasObject
+	slider  *Slider
+}
+
+// ApplyTheme is called when the Slider may need to update its look
+func (s *sliderRenderer) ApplyTheme() {
+	s.track.FillColor = theme.ButtonColor()
+	s.thumb.FillColor = theme.TextColor()
+	s.active.FillColor = theme.TextColor()
+	s.Refresh()
+}
+
+// Refresh is used to update the widget state for drawing
+func (s *sliderRenderer) Refresh() {
+	s.Layout(s.slider.Size())
+	canvas.Refresh(s.slider)
+}
+
+// Layout the components of the slider widget
+func (s *sliderRenderer) Layout(size fyne.Size) {
+	padLen := theme.Padding()
+	sideLen := padLen * standardScale
+	activeOffset, thumbOffset := s.moveSlide(sideLen)
+
+	var trackPos, activePos, thumbPos fyne.Position
+	var trackSize, activeSize fyne.Size
+
+	if s.slider.Vertical {
+		trackPos = fyne.NewPos(size.Width/2, 0)
+		activePos = fyne.NewPos(trackPos.X, activeOffset)
+
+		trackSize = fyne.NewSize(padLen, size.Height)
+		activeSize = fyne.NewSize(padLen, trackSize.Height-activeOffset)
+
+		thumbPos = fyne.NewPos(
+			trackPos.X-(sideLen-trackSize.Width)/2, thumbOffset)
+	} else {
+		trackPos = fyne.NewPos(0, size.Height/2)
+		activePos = trackPos
+
+		trackSize = fyne.NewSize(size.Width, padLen)
+		activeSize = fyne.NewSize(activeOffset, padLen)
+
+		thumbPos = fyne.NewPos(
+			thumbOffset, trackPos.Y-(sideLen-trackSize.Height)/2)
+	}
+
+	s.track.Move(trackPos)
+	s.track.Resize(trackSize)
+
+	s.active.Move(activePos)
+	s.active.Resize(activeSize)
+
+	s.thumb.Move(thumbPos)
+	s.thumb.Resize(fyne.NewSize(sideLen, sideLen))
+}
+
+// MinSize calculates the minimum size of a slider widget
+func (s *sliderRenderer) MinSize() fyne.Size {
+	s1, s2 := minLongSide, theme.Padding()*standardScale
+	if s.slider.Vertical {
+		return fyne.NewSize(s2, s1)
+	}
+	return fyne.NewSize(s1, s2)
+}
+
+func (s *sliderRenderer) BackgroundColor() color.Color {
+	return theme.BackgroundColor()
+}
+
+func (s *sliderRenderer) Destroy() {
+}
+
+func (s *sliderRenderer) Objects() []fyne.CanvasObject {
+	return s.objects
+}
+
+func (s *sliderRenderer) moveSlide(diameter int) (int, int) {
+	w := s.slider
+	t := s.track.Size()
+	ratio := (w.Value - w.Min) / (w.Max - w.Min)
+	if w.Vertical {
+		y := float64(t.Height) - (ratio * float64(t.Height))
+		return int(y), int(y) - int((1-ratio)*float64(diameter))
+	}
+	x := ratio * float64(t.Width)
+	return int(x), int(x) - int(ratio*float64(diameter))
+}
+
+func checkStep(s, max, min float64) {
+	// make sure there is a positive step and it is less than
+	// the maximum value
+	if s <= 0 {
+		fyne.LogError("Step is less than or equal to zero.", nil)
+	}
+	if s*s >= (max-min)*(max-min) {
+		fyne.LogError("Step is greater than or equal to range.", nil)
+	}
+}
+
+func checkMinMax(val, min, max float64) (float64, float64) {
+	// sort the values to ensure correct order
+	if val < min {
+		fyne.LogError("Value is less minimum value.", nil)
+		min = val
```

Comment:
I don't think we should re-order values if they are incorrect. The error should be sufficient.
It seems like the reason for this is due to how many values are passed to the constructor function, more in the comment below...

Reply:
I beg to differ about the reason. I am not saying that I am against the passing of less values to the constructor, but the real reason that its a problem is because the user **can** do it. 

Assessments should not made based on what you (the designer) thinks the users intentions are, but on their capabilities. I personally agree that the re-ordering of values should not be there, it makes the code uglier. 

Personally, I think the idea of having the error statements or sanitizing functions is effectively useless, since, as you mention below a user could just make a struct without using the constructor. Furthermore, all of the fields are public so even if you created a setter that you want run those sanitize functions inside of, the user could just access the fields directly.   

I understand the design justification behind the public fields. If you want to marshal in the values say from JSON, the field must be public. I get the reasoning behind it, but it does act counter encapsulation. I use that in the loosest meaning I know go isn't technically OO.    

Further thoughts below.....

Reply:
I agree with your statement about the errors could be bypassed - perhaps these should be checked outside the constructor function instead? I agree that we should not make assumptions about the users of the API - I think that backs up the idea of not changing parameters automatically for them. They have introduced a bug which should be fixed rather than automatically accommodated...